compat/src/suspense.js

```diff
@@ -61,40 +53,32 @@ Suspense.prototype = new Component();
 Suspense.prototype._childDidSuspend = function(promise) {
 	/** @type {import('./internal').SuspenseComponent} */
 	const c = this;
-	c._suspensions.push(promise);
 
 	const onSuspensionComplete = () => {
-		// From https://twitter.com/Rich_Harris/status/1125850391155965952
-		c._suspensions[c._suspensions.indexOf(promise)] =
-			c._suspensions[c._suspensions.length - 1];
-		c._suspensions.pop();
-
-		if (c._suspensions.length == 0) {
-			// If fallback is null, don't try to unmount it
-			// `unmount` expects a real VNode, not null values
-			if (c._fallback) {
-				// Unmount current children (should be fallback)
-				unmount(c._fallback);
-			}
-			c._vnode._dom = null;
-
-			c._vnode._children = c.state._parkedChildren;
-			c.setState({ _parkedChildren: null });
+		if (!--c._suspensions) {
+			c._detachOnNextRender = false;
+			c._vnode._children[0]._children = c.state._suspended;
+			c.setState({ _suspended: null });
 		}
 	};
 
-	if (c.state._parkedChildren == null) {
-		c._fallback = c._fallback && cloneElement(c._fallback);
-		c.setState({ _parkedChildren: c._vnode._children });
-		detachDom(c._vnode._children);
-		c._vnode._children = [];
+	if (!c._suspensions++) {
```

Comment:
I somewhen did an effort to de-duplicate `c._suspensions` when throwing the same promise multiple times which lead to duplicates in the `_suspensions` array and therefore to never un-suspend. 

Honestly I do not remember where this effort went as it seems to not be present in the current implementation.

We should check that this refactor still works when throwing the same promise multiple times.

Reply:
Yep, that's a good test. I'll add it to the suite.

I _suspect_ that the original array-based implementation (and this modified counter-based implementation) will survive the scenario even without special dedup logic. If the same promise is thrown from two children then it's added to the array twice (or the counter is incremented twice in the new implementation). When the promise resolves it's removed from the array twice (or the counter in decremented twice in the new implementation), which keeps the internal state consistent.

Reply:
I added the test, and it seems that both the original implementation and this modified one survive it just fine 👍